jose/constants.py

```diff
@@ -37,3 +41,15 @@ class Algorithms(object):
 
 
 ALGORITHMS = Algorithms()
+
+
+class Usages(object):
+    PUBLIC = 'public'
```

Comment:
Looking at this again I was a bit puzzled why the need to specify usage. Turns out that both Ed25519 public and private keys are 32 bytes in length and there's no clear way to distinguish them. It seems that an extra parameter is a clumsy way to indicate the key type, with a potential for security issues, eg. by mixing up key usage.

Luckily, it also turns out that a PEM key format for Ed25519 keys was standardized in [RFC8410](https://tools.ietf.org/html/rfc8410) while this PR was open. Thus, I think a self describing format would make it harder to use keys incorrectly and would also make this implementation much more maintainable in the long run by not having to provide backwards compatibility by accepting raw keydata arguments.

Additionally, I think it would be beneficial to merge the mpdavis:backend-explicit-tests branch into master before merging this since there's been a lot of work done to make the backends independent and this is yet another backend. That branch also makes python-jose depend directly on pyasn1, which would be needed to implement PEM (de)serialization for this backend.

Reply:
Good catch! I’m a little nervous that it’s still considered a draft specification, but I like the fact that our users would not need to explicitly specify the usage.

Your merge order seems sound to me, I’ll let you handle that merge since you’re already familiar with it and then I’ll deal with the rebase/merge fallout from that and implement reading from PEM files.

I’m not super happy about merging support for a specification that is in draft status, but I’ll try to keep an eye on that specification and keep this PR up to date with it. I’m also not sure how much proposed IETF standards change from draft to approval, and the only errata that has been reported so far is incredibly trivial. If that means we cut a release without this merged in, that’s alright.

Reply:
@blag @zejn

I finally had some time to sit down with this, and I wanted to try and get some clarity on what exactly the ask is here.

Is there anything about JOSE that needs the RFC8410 encoding specifically, or are you just looking for *some* self-describing format? Looking at the way that the keys are loaded from the JOSE payload, it looks like they are only ever expected in "raw" format. Is this correct?

If you are just looking for *a* self-describing format and don't specifically need RFC8410, then especially since RFC8410 has yet to be finalized, I would like to propose an alternate approach:

As of v2.6, `pyca/cryptography` supports Ed25519 signing and verification[1]. It looks like this PR was originally cut before that was released (2019-02-27). While pyca/cryptography does not support RFC8410 encoding, and likely will not until OpenSSL does, it does support PKCS8 encoding of Ed25519 keys, which does give you the self-describing nature that you are looking for.

Replacing the `PyNaCl` use in this backend with `pyca/cryptography` use should be pretty straightforward, and that would let us avoid adding another leaf to the already complex dependency graph.

As for the explicit "use" flag in the API, a simpler approach might be to have the constructor expect an already loaded key and have separate `classmethod`s for loading a private key vs public key from "raw" encoded bytes.

[1] https://cryptography.io/en/latest/changelog/#v2-6

Reply:
Any thoughts on this @blag @zejn @mpdavis ?

Reply:
Thanks for digging into this @mattsb42-aws, I really appreciate it.

I think I'll close this PR and release a new version of python-jose later this week from `master`. That way we at least get `@zejn`'s patches out the door.

Once that's done I'll dig into getting Ed25519 support with the Cryptography library.

How does that sound?

Reply:
Hmmm, Cryptography pretty much only offers Ed25519 (and some others for EdDSA), but no X25519 or other crypto primitives that could be used for actually enabling a "JOSE" library to do any more than JWS/JWT (e.g. using X25519 for ECDH for JWE).

So if someone was to asked me, my **very strong** preference for addressing this problem is still to provide a PyNaCl back-end, which would (down the line) resolve many more issues and enable further extensions not quite possible with Cryptography only (e.g. also modern and well accepted ciphers such as ChaCha20/Poly1305).

Just my NZ$0.02 on that topic ...

Reply:
Addendum: I've just seen that Cryptography may have pulled even with providing ChaCha20/Poly1305 and X25519. So from that angle it would be OK to use Cryptography instead of PyNaCl.

My only preference for PyNaCl resides in the fact that its implementations as well as bindings to `libsodium` have been long tested and found to be good and sufficiently tested and performant.

Reply:
I don't really have a stake in this, but from my perspective I'd be happy with either direction. PyNaCl and pyca/cryptography are owned and maintained by mostly the same people, and libsodium and libcrypto are both well established cryptographic codebases.

The main reason I would argue for using pyca/cryptography is to reduce the number of backends.

Reply:
The problem is not the source/author of the package (`pyca/cryptography` vs. `pyca/pynacl`), it's the source of the implementation of the ciphers. Cryptography used OpenSSL, which is known to be quite problematic. Its code base is not very good, thus the reasons for all the forks (e.g. LibreSSL, BoringSSL and others), particularly after the Heartbleed "incident".

In contrast libsodium was **always** very tight, secure and tidy.

From a functionality perspective though, I'd say you can go with Cryptography ...